File: src/platform_message/wlr_layer.dart
import "binary.dart";

enum Layer {
  background,
  bottom,
  top,
  overlay,
}

class Anchor {
  final bool top;
  final bool bottom;
  final bool left;
  final bool right;

  const Anchor({
    this.top = false,
    this.bottom = false,
    this.left = false,
    this.right = false,
  });
}

int serializeLayer(Layer layer) => switch (layer) {
      Layer.background => 0,
      Layer.bottom => 1,
      Layer.top => 2,
      Layer.overlay => 3,
    };

int serializeAnchor(Anchor anchor) =>
    (1 * (anchor.top as int)) |
    (2 * (anchor.bottom as int)) |
    (4 * (anchor.left as int)) |
    (8 * (anchor.right as int));

mixin WlrLayerBinding {
  static Map<int, WlrLayerBinding> instances = {};
}

Future<int> create(
  Layer layer,
  Anchor anchor,
  String namespace,
  WlrLayerBinding binding,
) async {
  final response =
      await sendPlatformMessage("wayland/wlr_layer/create", (writer) {
    writer.writeUint8(serializeLayer(layer));
    writer.writeUint8(serializeAnchor(anchor));
    writer.writeUtf8(namespace);
  });

  final viewId = response.readInt64();
  response.assertFinished();

  WlrLayerBinding.instances[viewId] = binding;

  // await initialCommit(viewId);
  return viewId;
}

Future<void> update(
  int viewId,
  int width,
  int height,
  Layer layer,
  Anchor anchor,
) async {
  await sendPlatformMessage("wayland/wlr_layer/update", (writer) {
    writer.writeInt64(viewId);
    writer.writeUint32(width);
    writer.writeUint32(height);
    writer.writeUint8(serializeLayer(layer));
    writer.writeUint8(serializeAnchor(anchor));
  });
}

Future<void> remove(int viewId) async {
  await sendPlatformMessage("wayland/wlr_layer/remove", (writer) {
    writer.writeInt64(viewId);
  });
  WlrLayerBinding.instances.remove(viewId);
}

File: src/platform_message/wlr_layer.rs
use std::{
    ffi::CStr,
    io::{Read, Result, Seek, Write},
};

use volito::ViewId;
use tracing::{error, info};

use crate::{
    embedder::FlutterWaylandSurface,
    nelly::{Nelly, NellyEvent},
    platform_message::ViewIdCounter,
    shell::layer::{Anchor, Layer},
};

use super::binary::{BinaryDecodable, BinaryReader, BinaryWriter};

impl BinaryDecodable for Layer {
    fn decode(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self> {
        match reader.read::<u8>()? {
            0 => Ok(Self::Background),
            1 => Ok(Self::Bottom),
            2 => Ok(Self::Top),
            3 => Ok(Self::Overlay),
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "invalid layer value",
            )),
        }
    }
}

impl BinaryDecodable for Anchor {
    fn decode(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self> {
        Anchor::from_bits(reader.read::<u32>()?).ok_or(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "invalid anchor value",
        ))
    }
}

#[derive(Debug)]
pub struct Create {
    layer: Layer,
    namespace: String,
}

impl BinaryDecodable for Create {
    fn decode(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self> {
        Ok(Self {
            layer: reader.read()?,
            namespace: reader.read_string()?,
        })
    }
}

impl super::PlatformRequest for Create {
    const CHANNEL: &'static CStr = c"wayland/wlr_layer/create";

    fn run(self, nelly: &mut Nelly, writer: &mut BinaryWriter<impl Write>) -> Result<()> {
        static VIEW_ID: ViewIdCounter = ViewIdCounter::new();
        let view_id = VIEW_ID.next_view_id();

        let surface = nelly.compositor_state.create_surface(&nelly.qh, view_id);

        let layer = nelly.layer_shell.create_layer_surface(
            &nelly.qh,
            surface,
            self.layer,
            self.namespace,
            None,
        );

        nelly
            .views
            .lock()
            .unwrap()
            .insert(view_id, FlutterWaylandSurface::from(layer));

        writer.write::<i64>(&view_id.0)
    }
}

#[derive(Debug)]
pub struct Update {
    view_id: ViewId,

    width: u32,
    height: u32,

    anchor: Anchor,
}

impl BinaryDecodable for Update {
    fn decode(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self> {
        Ok(Self {
            view_id: reader.read()?,
            width: reader.read()?,
            height: reader.read()?,
            anchor: reader.read()?,
        })
    }
}

impl super::PlatformRequest for Update {
    const CHANNEL: &'static CStr = c"wayland/wlr_layer/update";

    fn run(self, nelly: &mut Nelly, _writer: &mut BinaryWriter<impl Write>) -> Result<()> {
        let views = nelly.views.lock().unwrap();
        let window = views
            .get(&self.view_id)
            .and_then(|surface| {
                if let FlutterWaylandSurface::WlrLayer(surface) = surface {
                    Some(surface)
                } else {
                    None
                }
            })
            .expect("wlr_layer/update: view_id not found");

        window.set_size(self.width, self.height);
        window.set_anchor(self.anchor);

        Ok(())
    }
}

#[derive(Debug)]
pub struct Remove {
    view_id: ViewId,
}

impl BinaryDecodable for Remove {
    fn decode(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self> {
        Ok(Self {
            view_id: reader.read()?,
        })
    }
}

impl super::PlatformRequest for Remove {
    const CHANNEL: &'static CStr = c"wayland/wlr_layer/remove";

    fn run(self, nelly: &mut Nelly, _writer: &mut BinaryWriter<impl Write>) -> Result<()> {
        nelly.remove_view(self.view_id)?;

        Ok(())
    }
}

// pub struct CloseXdgToplevel {
//     pub view_id: ViewId,
// }

// impl super::ManagedPlatformEvent for CloseXdgToplevel {
//     const CHANNEL: &'static CStr = c"wayland/wlr_layer/close";

//     type Response = ();

//     fn encode(&self, writer: &mut BinaryWriter<impl Write>) -> Result<()> {
//         writer.write::<i64>(&self.view_id.0)
//     }

//     fn decode_response(reader: &mut BinaryReader<impl Read + Seek>) -> Result<Self::Response> {
//         reader.assert_finished()
//     }
// }
